workload/tpch: Disable FMA to fix ccl/workload/allccl test on s390x

ruixin-bao opened the pull request:
Previously, TestDeterministicInitialData/tpch under ccl/workloadccl/allccl
fail on s390x due to mismatched checksum of tpch initial data. After
comparing the data between two different platforms (intel and s390x),
it was found that the mismatch is due to a difference in the calculated
value of totalPrice under tpchOrders.

The number difference is due to FMA (fused multiply and add) instruction,
and adding a float32() will disable the usage of FMA to fix this difference.
Through testing, this change does not affect the functionality of the
same test on intel. In addition, this patch could be helpful for other
non amd64 platforms that have FMA instruction available.

For FMA related information, it can be found at
https://golang.org/ref/spec#Floating_point_operators

Commit:
workload/tpch: Disable FMA to fix ccl/workload/allccl test on s390x

Previously, TestDeterministicInitialData/tpch under ccl/workloadccl/allccl
fail on s390x due to mismatched checksum of tpch initial data. After
comparing the data between two different platforms (intel and s390x),
it was found that the mismatch is due to a difference in the calculated
value of totalPrice under tpchOrders.

The number difference is due to FMA (fused multiply and add) instruction,
and adding a float32() will disable the usage of FMA to fix this difference.
Through testing, this change does not affect the functionality of the
same test on intel. In addition, this patch could be helpful for other
non amd64 platforms that have FMA instruction available.

For FMA related information, it can be found at
https://golang.org/ref/spec#Floating_point_operators

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54946) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=54946) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=54946) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54946)
<!-- Reviewable:end -->

ruixin-bao commented:
Hi,

It looks like the CI results showed two lint errors. Sorry, I rechecked the docs today and found I should run `make pre-push` command first before opening a PR (I only ran `make test` locally).

Seems like the error is due to this conversion of `float32( )`, which is unfortunate. Any suggestions on how to fix this lint error while being able to disable FMA? Thanks for your help.

yuzefovich reviewed (commented):
@ruixin-bao thanks for taking a look at this very peculiar issue!

There are two errors on the CI, one of them (`TestSchemaChangeEvalContext`) is a flake (and has either been skipped or fixed), so you just need to rebase on top of the latest master, and I left a comment below how to address another one.

You shouldn't worry about the CI being red though - that's what it is there for :)

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54946#-:-MIWNLm065FOsG9BmSdm:b-zcmshp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ruixin-bao)

___
*[pkg/workload/tpch/generate.go, line 406 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54946#-MIWNfyo3hSqIj-hpp3n:-MIWNfyo3hSqIj-hpp3o:b-8wdozf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d714c3d341f5159618fa44f1bc652dd0fde4b7bf/pkg/workload/tpch/generate.go#L406)):*
> ```Go
> 		for j := 0; j < l.orderData.nOrders; j++ {
> 			ep := l.orderData.quantities[j] * makeRetailPriceFromPartKey(l.orderData.partKeys[j])
> 			totalPrice += float32(ep * (1 + l.orderData.tax[j]) * (1 - l.orderData.discount[j]))
> ```

Our linter is complaining about unnecessary conversion, but I think you can go around that by adding
`//nolint:roachvet` to the end of the line.

Also, this will definitely need a comment describing why we're doing this what-seems-like-unnecessary conversion.


<!-- Sent from Reviewable.io -->

ruixin-bao commented:
Hi,
@yuzefovich, thanks for the review and the pointers : ).

> There are two errors on the CI, one of them (TestSchemaChangeEvalContext) is a flake (and has either been skipped or fixed), so you just need to rebase on top of the latest master, and I left a comment below how to address another one.

I see, I have rebased this patch on top of the master. 

> Our linter is complaining about unnecessary conversion, but I think you can go around that by adding
//nolint:roachvet to the end of the line.

I have tried the approach, seems like  `//nolint:roachvet` has not been added yet?  From this direction, I find some other lint comments include: `lint:ignore`, `nolint:returnerrcheck` and `nolint:fmtsafe`. Wondering if those would help? 

> Also, this will definitely need a comment describing why we're doing this what-seems-like-unnecessary conversion.

Agreed, added a comment in the new patch.

yuzefovich reviewed (commented):
Hm, I thought `nolint:roachvet` would just work, but now that I look into it a bit closer it makes sense that it wouldn't :/ `lint:ignore` sounds promising (but I'm guessing it still won't work), do you wanna give it a try? Two other directives are irrelevant.

@ajwerner @knz since you guys modified `TestRoachVet` last, do you have an idea of how to go around this linter issue? We could always make the file be filtered out, but I'm wondering whether there is a better approach.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54946#-:-MIe3y6-9vehaJG27QRU:bekp0cg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ruixin-bao)

___
*[pkg/workload/tpch/generate.go, line 407 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54946#-MIe5AJp7Qto9_aJ2NFg:-MIe5AJp7Qto9_aJ2NFh:biwxhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b35aceae3f8416dc49f88a5526d2495d252a68a2/pkg/workload/tpch/generate.go#L407)):*
> ```Go
> 			ep := l.orderData.quantities[j] * makeRetailPriceFromPartKey(l.orderData.partKeys[j])
> 			// Use an extra float32 conversion to disable "fused multiply and add" (FMA) to force
> 			// identical behaivor on all platforms. See https://golang.org/ref/spec#Floating_point_operators
> ```

nit: `s/behaivor/behavior/`.


<!-- Sent from Reviewable.io -->

ruixin-bao commented:
Hi,

> Hm, I thought nolint:roachvet would just work, but now that I look into it a bit closer it makes sense that it wouldn't :/ lint:ignore sounds promising (but I'm guessing it still won't work), do you wanna give it a try?

Yes, you are correct. I tried `lint:ignore` locally and looks like it did not work.

> @ajwerner @knz since you guys modified TestRoachVet last, do you have an idea of how to go around this linter issue? We could always make the file be filtered out, but I'm wondering whether there is a better approach.

Thanks,  I will wait a bit for this reply and update the PR accordingly. 

> nit: s/behaivor/behavior/.

Oops, thanks for catching that, will update it next time I update this PR.

ajwerner commented:
You're right that we should be able to defeat the linter with a comment but we currently can't. I started typing up the code to fix that but ran into some annoying issues with cases where the comment is on top of an if conditional like:

```
// nolint:... 
if _ = <lint problem>; ... {
}
```

ajwerner commented:
I'll try to finish up the fix tomorrow

ajwerner commented:
Posted https://github.com/cockroachdb/cockroach/pull/55218 to add support for `nolint:unconvert` comments.

ajwerner commented:
Alright, if you rename master, you should be able to add a `nolint:unconvert` comment to defeat the linter.

ruixin-bao commented:
@ajwerner Thank you for your PR. Much appreciated. I have rebased this patch to use `nolint:unconvert` and the lint tests did pass locally :).

yuzefovich reviewed (approved):
Thanks Andrew, and thanks @ruixin-bao for your contribution!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

bors r=yuzefovich

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54946#-:-MIzLbt39LHOyPq5UfMm:bkx44hp)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ruixin-bao commented:
Thank you for the reviews!

SWDevAngel commented:
@ruixin-bao Hi Ruixin. Thank you for contributing to CockroachDB this year. As a token of our appreciation we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) `@lisa` so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please send this asap. :)

ruixin-bao commented:
@SWDevAngel, thank you for this kind message. I will select the charitable contribution choice : ).

SWDevAngel commented:
Okay thanks @ruixin-bao. Can you please DM me on Slack or let me know where I can DM or email you a link so you can chose your charity? :)

ruixin-bao commented:
@SWDevAngel  Thanks, I messaged on slack.
